### PR TITLE
Make dht_put_alert more accurate.

### PR DESCRIPTION
**Pull request by thomas-yuan**
When I start using libtorrent DHT at beginning,  I didn't know there is a dht_put.cpp in tools. My colleague found a similar code and we using it as our test app. But seems like 'mput' doesn't work properly. After investigating for a while, I found the reason: the sign_mutable_item used a incorrect private key to sign the data, so actually DHT always get "invalid signature" error, but since the dht_put_alert was sent before 'put' and there is no indicator to say the 'put' is success or failed, so as a user, I wouldn't know the put actually was failed. instead, I always get dht_put_alert which seems like the 'put' is done successfully. 

That's why I created this PR.
Tested with dht_put.cpp, and seems work fine for me.

### CODE REVIEW

**Comment by thomas-yuan** [comment]
working on dht_test now.

File: include/libtorrent/alert_types.hpp [hunk 1]
```diff
@@ -2064,6 +2065,8 @@ namespace libtorrent
 		boost::array<char, 64> signature;
 		std::string salt;
 		boost::uint64_t seq;
+
+		bool success;
```
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: could you please add a comment describing what this means? (if this is false, does it mean it definitely failed, or may there be another put that succeeds later?)

File: include/libtorrent/kademlia/dht_tracker.hpp [hunk 2]
```diff
@@ -100,10 +100,11 @@ namespace libtorrent { namespace dht
 			, std::string salt = std::string());
 
 		void put_item(entry data
-			, boost::function<void()> cb);
+			, boost::function<void(bool)> cb);
```
Comment (arvidn) [inline include/libtorrent/kademlia/dht_tracker.hpp]: please add a comment describing what bool means here

File: include/libtorrent/kademlia/dht_tracker.hpp [hunk 3]
```diff
@@ -100,10 +100,11 @@ namespace libtorrent { namespace dht
...
 
 		void put_item(char const* key
-			, boost::function<void(item&)> cb, std::string salt = std::string());
+			, boost::function<void(item&, bool)> cb
```
Comment (arvidn) [inline include/libtorrent/kademlia/dht_tracker.hpp]: and maybe here too, unless it's obvious from a comment above

File: include/libtorrent/kademlia/get_item.hpp [hunk 4]
```diff
@@ -43,7 +43,7 @@ namespace libtorrent { namespace dht
 class get_item : public find_data
 {
 public:
-	typedef boost::function<bool(item&, bool)> data_callback;
+	typedef boost::function<void(item&, bool)> data_callback;
```
Comment (arvidn) [inline include/libtorrent/kademlia/get_item.hpp]: what happened to the return value here?

File: include/libtorrent/kademlia/put_data.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef PUT_DATA_150921_HPP
+#define PUT_DATA_150921_HPP
```
Comment (arvidn) [inline include/libtorrent/kademlia/put_data.hpp]: please use the same convention for include guards as other headers

File: src/kademlia/dht_tracker.cpp [hunk 6]
```diff
@@ -238,53 +238,10 @@ namespace libtorrent { namespace dht
 		m_dht.announce(ih, listen_port, flags, f);
 	}
 
-	namespace {
-
-	// these functions provide a slightly higher level
-	// interface to the get/put functionality in the DHT
-	bool get_immutable_item_callback(item& it, boost::function<void(item const&)> f)
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: what happened to all of these functions?

File: src/kademlia/get_item.cpp [hunk 7]
```diff
@@ -170,92 +162,28 @@ bool get_item::invoke(observer_ptr o)
 
 void get_item::done()
 {
+	// only for put_immutable_item.
+	if (!m_data_callback) return find_data::done();
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: is m_data_callback guaranteed to not be set for immutable puts? is it guaranteed to _be_ set for mutable puts?
it may be appropriate to put an assert here in either case (or expand the comment a bit)

File: src/kademlia/get_item.cpp [hunk 8]
```diff
@@ -170,92 +162,28 @@ bool get_item::invoke(observer_ptr o)
...
+
 	if (m_data.is_mutable() || m_data.empty())
 	{
 		// for mutable data, now we have authoritative data since
 		// we've heard from everyone, to be sure we got the
 		// latest version of the data (i.e. highest sequence number)
-		bool put_requested = m_data_callback(m_data, true);
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: did this code move into put_item.cpp?

File: src/kademlia/get_item.cpp [hunk 9]
```diff
@@ -170,92 +162,28 @@ bool get_item::invoke(observer_ptr o)
 
 void get_item::done()
 {
+	// only for put_immutable_item.
+	if (!m_data_callback) return find_data::done();
+
 	if (m_data.is_mutable() || m_data.empty())
 	{
 		// for mutable data, now we have authoritative data since
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: what happened to authoritative?

File: src/kademlia/node.cpp [hunk 10]
```diff
@@ -436,12 +435,12 @@ void node::get_item(sha1_hash const& target
 #endif
 
 	boost::intrusive_ptr<dht::get_item> ta;
-	ta.reset(new dht::get_item(*this, target, f));
+	ta.reset(new dht::get_item(*this, target, boost::bind(f, _1), find_data::nodes_callback()));
```
Comment (arvidn) [inline src/kademlia/node.cpp]: while you're here, would you mind using boost::make_shared<> instead?

File: src/kademlia/node.cpp [hunk 11]
```diff
@@ -453,7 +452,77 @@ void node::get_item(char const* pk, std::string const& salt
...
+namespace {
+
+void put(std::vector<std::pair<node_entry, std::string> > const& nodes
+	, boost::intrusive_ptr<dht::put_data> ta)
+{
+    ta->set_targets(nodes);
+	ta->start();
+}
+
+void put_data_cb(item& i, bool auth
```
Comment (arvidn) [inline src/kademlia/node.cpp]: I guess authoritative is still here.

File: src/kademlia/rpc_manager.cpp [hunk 12]
```diff
@@ -158,7 +159,7 @@ void observer::set_id(node_id const& id)
 enum { observer_size = max3<
 	sizeof(find_data_observer)
 	, sizeof(announce_observer)
-	, sizeof(null_observer)
+	, sizeof(put_data_observer)
```
Comment (arvidn) [inline src/kademlia/rpc_manager.cpp]: I think you need the null_observer in here too (for completeness). (extending max to take one more argument should be pretty straight-forward, or you could just combine multiple max3). or do you think there's no point?

**Comment by arvidn** [comment]
I haven't looked over and understood all yet. in general I would like to see more comments, to make it clearer what's going on

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (ssiloti) [inline src/kademlia/dht_tracker.cpp]: I'm relying on these wrappers to implement IPv4/IPv6 dual stack operations so they had better stick around.

File: include/libtorrent/alert_types.hpp (hunk 1 above)
Comment (thomas-yuan) [inline include/libtorrent/alert_types.hpp]: Will add comments for sure. actually I also not sure when we should set it as false.

I added some logs for testing, what I saw before is, sometimes some node response a error like "invalid token", sometimes is "unknown request" (old version I guess?), and if I use a wrong private key, all nodes reply "invalid signature".

Right now, only when (success response >= error responses), I set success to true, but i'm not sure if it's the best choice here.  (BTW, there is no another put later)

File: include/libtorrent/kademlia/get_item.hpp (hunk 4 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/get_item.hpp]: we need this value just because we mixed put/get together to get. now we have nodes_callback used for put, get_items will always just try to get. so we needn't the return value anymore.

File: include/libtorrent/kademlia/put_data.hpp (hunk 5 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/put_data.hpp]: This one is copied from find_data.hpp.  which one I should follow? just remove the "_150921" ?

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (thomas-yuan) [inline src/kademlia/dht_tracker.cpp]: @ssiloti The comments said we need this wrapper to control the return value. I didn't see any IP related stuff here.

File: src/kademlia/get_item.cpp (hunk 8 above)
Comment (thomas-yuan) [inline src/kademlia/get_item.cpp]: yes.

File: src/kademlia/rpc_manager.cpp (hunk 12 above)
Comment (thomas-yuan) [inline src/kademlia/rpc_manager.cpp]: ok, then actually we need get_item_observer, get_peers_observer, traversal_observer as well? 

Actually I guess the size should be same, only put_data_observer is bigger than others since it keep a string for token. 

But generally, we should put all observers here I think.

File: src/kademlia/node.cpp (hunk 10 above)
Comment (thomas-yuan) [inline src/kademlia/node.cpp]: like boost::shared_ptr ta = boost::make_shared<dht::get_item>(*this, target, f) ?

Usually i use std::make_shared (with std=c++11 for sure, and I know libtorrent also support c98 so we couldn't use it), I don't have any experience about boost::intrusive_ptr. But in kademlia, seems all codes are using it, not shared_ptr.

File: src/kademlia/get_item.cpp (hunk 7 above)
Comment (thomas-yuan) [inline src/kademlia/get_item.cpp]: Yes and Yes. 
For immutable_put, we only need nodes_callback then we can write item to them.
For mutable_put, we need data_callback to update the sequence number, so it must be set.

**Comment by thomas-yuan** [comment]
Right now, when we try to put something, we create get_item, and pass in a callback function. the return value of the callback function will indicate this is a put or get.

This change is kinda similar as get_peers/announce. 
- add nodes_callback to get_item, so when we try to put something, we need set nodes_callback. we needn't set data_callback for immutable, but for mutable, since we need sequence number, so we need a data_callback function. The data_callback will only be called when we get authoritative response, and the callback itself needn't authoritative parameters.
- generally we should create a put_data object and start it in nodes_callback. but for mutable item, the value is set in data_callback and I need a way to keep it until nodes_callback is called. so I create a put_data traversal_algorithm object at beginning and pass it around.
- put_data actually just invokes 'put' operations to all nodes(k = 8 I believe) we get from get_item. There is no travel at all, only collects how many operations success and how many failed. If success > failed, the alert will say 'put success'. In my test, unless the signature is wrong(all failed), usually only 1 or 2 failed.

File: include/libtorrent/alert_types.hpp (hunk 1 above)
Comment (thomas-yuan) [inline include/libtorrent/alert_types.hpp]: or remote the bool success, just add 
```c++
uint8_t m_response;
uint8_t m_timeout;
```

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (ssiloti) [inline src/kademlia/dht_tracker.cpp]: There isn't anything right now, but I have changes in the works which will expand these functions.

Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: @thomas-yuan would you mind putting these back (even if being redundant) to not collide too badly with ssiloti's patch?

File: src/kademlia/node.cpp (hunk 10 above)
Comment (arvidn) [inline src/kademlia/node.cpp]: ah, I see. I thought I had transitioned everything to shared_ptr (and I missed that you said intrusive_ptr right above). never mind then

File: src/kademlia/rpc_manager.cpp (hunk 12 above)
Comment (arvidn) [inline src/kademlia/rpc_manager.cpp]: yeah, you're right. do you want me to do that as a separate patch?

**Comment by arvidn** [comment]
that makes sense. there's still a callback for receiving the (mutable) data that's called for *any* valid data we receive too, right? and then it's called one last time with the authoritative set once we exhausted the search. This is a critical optimization to be able to start processing data optimistically.

File: include/libtorrent/alert_types.hpp [hunk 13]
```diff
@@ -2064,6 +2065,12 @@ namespace libtorrent
 		boost::array<char, 64> signature;
 		std::string salt;
 		boost::uint64_t seq;
+
+		// DHT put operation usually writes item to k nodes, maybe the node
+		// is stale so no response, or the node doesn't support 'put', or the
+		// token for write is out of date, etc.
+		// set success to true when success responses >= failed responses.
```
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: does no response at all count as a "failed response"? This condition seems a bit arbitrary. wouldn't it make more sense to just include the number of confirmed successful responses instead?

File: include/libtorrent/kademlia/put_data.hpp [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef PUT_DATA_HPP
```
Comment (arvidn) [inline include/libtorrent/kademlia/put_data.hpp]: would you mind using the same naming convention as the other header files?

File: include/libtorrent/kademlia/get_item.hpp (hunk 4 above)
Comment (arvidn) [inline include/libtorrent/kademlia/get_item.hpp]: is nodes_callback called when we receive an authoritative response? and it triggers the put?
for immutable items, what's the distinction between data_callback and nodes_callback?

Comment (thomas-yuan) [inline include/libtorrent/kademlia/get_item.hpp]: nodes_callback will be called after we receive an authoritative response (get_item::done() will call data_callback first, then it call find_data::done() which will call nodes_callback). 

nodes_callback will trigger put.

for immutable items, get will only supply data_callback, and put will only supply nodes_callback.

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (thomas-yuan) [inline src/kademlia/dht_tracker.cpp]: ok.

File: src/kademlia/rpc_manager.cpp (hunk 12 above)
Comment (thomas-yuan) [inline src/kademlia/rpc_manager.cpp]: That will be better.

File: include/libtorrent/kademlia/put_data.hpp (hunk 14 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/put_data.hpp]: Sure, but I don't know which one you want.
in dht_storage.hpp
#ifndef DHT_STORAGE_HPP       (file name)

in dht_tracker.hpp
#ifndef TORRENT_DHT_TRACKER    (torrent + file name)

in dos_blocker.hpp
#ifndef TORRENT_DHT_DOS_BLOCKER    (torrent + dht + file name)

in find_data.hpp
#ifndef FIND_DATA_050323_HPP     (file name + date)

in get_item.hpp
#ifndef LIBTORRENT_GET_ITEM_HPP    (libtorrent + file name)

in node_entry.hpp
#ifndef KADEMLIA_NODE_ENTRY_HPP    (kademlia + file name)

File: include/libtorrent/alert_types.hpp (hunk 13 above)
Comment (thomas-yuan) [inline include/libtorrent/alert_types.hpp]: yes. 

Good idea, actually I remember I wrote similar idea (the number of confirmed successful and also the number of time out) here but I can't find it.

File: include/libtorrent/kademlia/put_data.hpp (hunk 14 above)
Comment (arvidn) [inline include/libtorrent/kademlia/put_data.hpp]: heh, I guess I haven't been paying attention to this lately :)
The ones used by the rest of libtorrent is TORRENT_*_HPP (and no date).
I already asked aldenml to change the dht_storage one

File: test/test_dht.cpp [hunk 15]
```diff
@@ -1788,10 +1801,10 @@ TORRENT_TEST(dht)
 		for (int i = 0; i < num_test_nodes; ++i)
 			node.m_table.add_node(nodes[i]);
 
-		sha1_hash target = hasher(public_key, item_pk_len).final();
+//		sha1_hash target = hasher(public_key, item_pk_len).final();
```
Comment (arvidn) [inline test/test_dht.cpp]: could you delete this line too?

File: test/test_dht.cpp [hunk 16]
```diff
@@ -1626,8 +1642,8 @@ TORRENT_TEST(dht)
 		udp::endpoint initial_node(address_v4::from_string("4.4.4.4"), 1234);
 		node.m_table.add_node(initial_node);
 
-		sha1_hash target = hasher(public_key, item_pk_len).final();
-		node.get_item(target, get_item_cb);
+//		sha1_hash target = hasher(public_key, item_pk_len).final();
```
Comment (arvidn) [inline test/test_dht.cpp]: could you delete this line?

File: src/kademlia/node.cpp [hunk 17]
```diff
@@ -453,7 +452,77 @@ void node::get_item(char const* pk, std::string const& salt
...
+		, boost::intrusive_ptr<put_data> ta
+		, boost::function<void(item&)> f)
+{
+	// call data_callback only when we got authoritative data.
```
Comment (arvidn) [inline src/kademlia/node.cpp]: this comment makes me suspicious though. The point of the authoritative flag is for the user to get multiple data callbacks, the first ones coming back very quickly, but without being authoritative and then the last one (which probably has to wait for a few node time-outs, and will be very late) being authoritative. i.e. not until the last one can one be somewhat certain that it is the latest version.

The benefit of this is that clients can start perform work optimistically, and discard it later if it turned out it wasn't the latest version of the data. But chances are that the first data blob you get back _is_ the latest, or at least contain some relevant data.

File: include/libtorrent/kademlia/get_item.hpp (hunk 4 above)
Comment (arvidn) [inline include/libtorrent/kademlia/get_item.hpp]: right. it makes sense that you don't trigger the put until you get an authoritative response. Traditionally though, the nodes_callback is called when we receive a response without any data (or am I misremembering that?). i.e. it just contains "nodes" to continue the traversion, but no "values" with bittorrent peers. I'm trying to understand why it makes sense to call this callback nodes_callback. It seems like it should be something like completed_callback (or something along those lines).

Comment (thomas-yuan) [inline include/libtorrent/kademlia/get_item.hpp]: oh, Then I have different understanding for nodes_callback. I think it's like a results not for traversal, but for put(just like completed_callback you presented here). All nodes in the callback must be the k closest nodes to the target.

So in my mind, it's just the completed_callback, the first step for put (announce/put) something to DHT. (The nodes for traversal logic is a internal logic, need call any callback)

File: src/kademlia/node.cpp (hunk 17 above)
Comment (thomas-yuan) [inline src/kademlia/node.cpp]: Yes, "the authoritative flag is for the user to get multiple data callbacks", that only used for "GET", so user can get data very quickly. but for "PUT", we must waiting the authoritative result because we need change the sequence number and sign it. 

The logic didn't changed, before this patch, the logic looks like this (cb is data_callback):
https://github.com/arvidn/libtorrent/blob/master/src/kademlia/dht_tracker.cpp#L275

```c++
	bool put_mutable_item_callback(item& it, bool authoritative, boost::function<void(item&)> cb)
	{
		if (authoritative) {
			cb(it);
		}
		return true;
	}
```

**Comment by thomas-yuan** [comment]
@arvidn I bring back the original callback functions, it breaks CI. do you have any suggestion for resolving the unused functions error? 

Also, I didn't merge patches because that will break the line comments. you can ask me to merge all of these when everything looks ok for you.

**Comment by thomas-yuan** [comment]
@ssiloti reviewed your dht_ipv6 branch. 

Looks like you need the callback functions when send query to both ipv4 and ipv6, to merge the results, so user will only get one callback for immutable items, and one authoritative callback for mutable items.

It's fine for immutable item since it's the hash of content, but for mutable item, maybe the results are different. 

For example, I have two devices, they're using the same key-pairs since they are all my devices. Device A maybe connected to both IPv4 and IPv6, but device B only connected to IPv4(or only IPv6). they all put data to DHT, so the data in IPv4 and IPv6 are different. If someone else query my public key, he can get both if he also connected to both IPv4/6, we can't make a decision which one is new or valuable for user, we problely need post all of them to user, but add a indicator saying where it comes from.

Thought?

**Comment by ssiloti** [comment]
@thomas-yuan that's a bug in get_mutable_item_callback. The code should look like this:

    if (authoritative && --ctx->active_traversals == 0)
    			f(it, authoritative);
    		else
    			f(it, false);

So the callback still gets called for all results but authoritative is not set until both traversals complete.

**Comment by thomas-yuan** [comment]
@ssiloti In my mind, authoritative means the result is the final result, which should overwrite all other results. My point is, maybe we should have two final results, one for IPv4, one for IPv6. Just let user make a choice, means user should know he can get two different results from IPv4 and IPv6 since he enabled IPv6. 
That doesn't make sense for a real user, but here 'user' means other developers who use this library.

**Comment by thomas-yuan** [comment]
failing caused by ssl_test.

**Comment by ssiloti** [comment]
IMO clients should not have to know or care whether they are getting results from IPv4, IPv6, or both. I expect that, at least initially, the vast majority of IPv6 nodes will also be running IPv4 so the data should be mostly coherent between the two anyways.

**Comment by thomas-yuan** [comment]
@arvidn anything else I need to do for this PR?

**Comment by thomas-yuan** [comment]
@arvidn merge all patches to one. Let me know what else I should do to getting this patch to be accepted.

**Comment by arvidn** [comment]
sorry for having put off reviewing this (it's a large patch, which makes it feel scarier). To some degree I was hoping that @ssiloti would comment on it too. I would like to avoid creating unnecessary conflicts with his patch, or make him have to revert anything.

**Comment by aldenml** [comment]
@thomas-yuan every time I review this patch, I feel like asking myself: why? what's the problem exactly?

I see no problem with the semantics of the current alert (just maybe the name is a little misleading). If you want to make sure you have succeed with the put, why just don't request it from the DHT?

**Comment by ssiloti** [comment]
The removable of the intermediate callbacks from dht_tracker is the only non-trivial overlap between this patch and my IPv6 changes. Even though we're touching a lot of the same files, we're mostly concerned with different layers of the DHT code.

In general I'm a bit concerned about the size of this patch though. There is a lot of code being added here for such a seemingly simple problem as the one cited at the top of this PR. I'll need to take a closer look before I can say for sure how much of it is really necessary.

**Comment by thomas-yuan** [comment]
@aldenml The problem is, when you get dht_put_alert, it only means you found nodes to put. (I guess that's your option, it's misleading, but it's acceptable.) Actually if you checking the code, you will find there are comments like this:
```
		// TODO: ideally this function would be called when the		
		// put completes
```
So basically, this patch is for that TODO. Also, If I'm a user of libtorrent, I would presume when I get dht_put_alert, the data has been successfully put to DHT. 

As you said, I can check it by send another request, but If we thinking DHT as storage like DB, could you imagining that when you write something to DB, the API return success but actually not? You need  query it to verify it?

**Comment by aldenml** [comment]
As far as I can see, there is a matter of design here (the exact meaning of `dht_put_alert` and a possible alternative). If you think it will help, I recommend the creation of an issue to discuss about it.

**Comment by ssiloti** [comment]
I've created #232 to address the specific problem identified in the opening of this PR. I'm not convinced we should report any error responses from remote nodes to the user. In general it's not clear what the client should do in response to these errors.

**Comment by arvidn** [comment]
I believe certain errors need to be reported to the user, or at least to the put-callback. Specifically the "cas-failure", or "someone replaced this blob in between you reading it and writing it back". I believe this is supported currently. I do agree that this patch does look bigger than it needs to be. Even if all changes are good, anything that doesn't directly solve the problem of posting the alert may be better put in a separate PR. One reason why I've procrastinated so long trying to understand everything going on here is partly because of the size of the patch.

File: include/libtorrent/alert_types.hpp [hunk 18]
```diff
@@ -2066,6 +2067,11 @@ namespace libtorrent
 		boost::array<char, 64> signature;
 		std::string salt;
 		boost::uint64_t seq;
+
+		// DHT put operation usually writes item to k nodes, maybe the node
+		// is stale so no response, or the node doesn't support 'put', or the
+		// token for write is out of date, etc.
```
Comment (arvidn) [inline include/libtorrent/alert_types.hpp]: I suppose this is supposed to count the number of successful responses we got from the puts. (but the comment doesn't actually spell that out). Btw, do you know if we keep trying more nodes, farther away from the target, if puts fail? to make a best-effort attempt at storing the blob on at least k nodes.

File: include/libtorrent/kademlia/dht_tracker.hpp [hunk 19]
```diff
@@ -101,11 +101,18 @@ namespace libtorrent { namespace dht
...
+		// for immutable_item.
+		// the callback function will be called when put operation is done.
+		// the bool parameter indicates the put operation success or failed.
 		void put_item(entry data
-			, boost::function<void()> cb);
+			, boost::function<void(int)> cb);
 
+		// for mutable_item.
+		// the data_cb will be called when we get authoritative mutable_item,
+		// the cb is same as put immutable_item.
```
Comment (arvidn) [inline include/libtorrent/kademlia/dht_tracker.hpp]: My recollection of the semantics of the callback here is that it's called with 'item' containing the latest version of the blob (meaning we have received the authoritative version). The user is expected to potentially modify the item and it will then be put back into the DHT. If the DHT has been updated in between and one or more puts fail with that error, we'll try again.

Is this your understanding too? (sorry, it's been a while since I've been working in this code).

File: include/libtorrent/kademlia/dht_tracker.hpp [hunk 20]
```diff
@@ -101,11 +101,18 @@ namespace libtorrent { namespace dht
 			, boost::function<void(item const&, bool)> cb
 			, std::string salt = std::string());
 
+		// for immutable_item.
+		// the callback function will be called when put operation is done.
+		// the bool parameter indicates the put operation success or failed.
```
Comment (arvidn) [inline include/libtorrent/kademlia/dht_tracker.hpp]: this comment refers to bool, but the parameter is int now

File: include/libtorrent/kademlia/get_item.hpp [hunk 21]
```diff
@@ -53,13 +53,15 @@ class get_item : public find_data
 	// for immutable itms
 	get_item(node& dht_node
 		, node_id target
-		, data_callback const& dcallback);
+		, data_callback const& dcallback
+		, nodes_callback const& ncallback);
```
Comment (arvidn) [inline include/libtorrent/kademlia/get_item.hpp]: I can't actually remember the exact semantics of these callbacks. When is this new callback called again? and would you mind adding a comment in the code?

Comment (arvidn) [inline include/libtorrent/kademlia/get_item.hpp]: allright, nodes_callback seems to be called when a find_data traversal completes. and it passes on the closest DHT nodes we found during the search, the ones we may want to put to.

File: src/kademlia/dht_tracker.cpp [hunk 22]
```diff
@@ -294,26 +251,25 @@ namespace libtorrent { namespace dht
...
+		, boost::function<void(int)> cb)
 	{
 		std::string flat_data;
 		bencode(std::back_inserter(flat_data), data);
 		sha1_hash target = item_target_id(
 			std::pair<char const*, int>(flat_data.c_str(), flat_data.size()));
 
-		m_dht.get_item(target, boost::bind(&put_immutable_item_callback
-			, _1, cb, data));
+		m_dht.put_item(target, data, cb);
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: it's not entirely obvious to me what the benefit of introducing put_item() and get_item() is, over having a get_item() where you can choose to also put once your callback is called. Is there state that's lost that can be more efficiently handed off across traversal_algorithms? It seems like this just makes the interface wider without simplifying the get_item() function. I suppose you can save one or two layers of forwarding boost::function calls.

File: src/kademlia/dht_tracker.cpp [hunk 23]
```diff
@@ -294,26 +251,25 @@ namespace libtorrent { namespace dht
...
 	}
 
 	void dht_tracker::put_item(char const* key
-		, boost::function<void(item&)> cb, std::string salt)
+		, boost::function<void(item&, int)> cb
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: granted, before this interface reaches the user, it should be simpler and probably separate get and put. In the intermediate layers the interfaces should probably be as narrow as possible (that mostly applies to the user-facing interface as well). Now, I think dht_tracker is probably entirely redundant too, and just cause overhead in passing calls back and forth.

File: src/kademlia/get_item.cpp [hunk 24]
```diff
@@ -51,6 +51,9 @@ void get_item::got_data(bdecode_node const& v,
 {
 	// we received data!
 
+	// for put_immutable_item, there is no data_callback.
+	if (!m_data_callback) return;
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: why do we need to skip this function here? did there used to be a data callback for put_immutable_item() or was this just an unrelated bug with no test coverage?

File: src/kademlia/get_item.cpp [hunk 25]
```diff
@@ -88,39 +91,27 @@ void get_item::got_data(bdecode_node const& v,
...
 
 		m_data.assign(v);
-		bool put_requested = m_data_callback(m_data, true);
 
-		// if we intend to put, we need to keep going
-		// until we find the closest nodes, since those
-		// are the ones we're putting to
-		if (put_requested)
-		{
+		// There can only be one true immutable item with a given id
```
Comment (arvidn) [inline src/kademlia/get_item.cpp]: is it a safe assumption that we're not putting a mutable blob here? just because I asked for a mutable blob doesn't mean the other end can't send me a response that looks like an immutable one. That shouldn't make me abort though, right? or am I misreading this?

It looks like we make an assumption about what _we_ are doing and want to do based on what an untrusted peer told us.

Another powerful feature that may be lost here is that the client can choose whether to update the blob or not _after_ having seen it. Perhaps it doesn't need any updating, it's already at the latest version. We won't know that until we read it from the DHT.

File: src/kademlia/node.cpp [hunk 26]
```diff
@@ -454,7 +453,77 @@ void node::get_item(char const* pk, std::string const& salt
...
+
+} // namespace
+
+void node::put_item(sha1_hash const& target, entry& data, boost::function<void(int)> f)
+{
+#ifndef TORRENT_DISABLE_LOGGING
+	if (m_observer)
+	{
+		char hex_target[41];
+		to_hex(reinterpret_cast<char const*>(&target[0]), 20, hex_target);
```
Comment (arvidn) [inline src/kademlia/node.cpp]: instead of reinterpret_cast here, you can use target.data() (iirc)

**Comment by arvidn** [comment]
apart from my comments, I can't find any obvious issues. Do you feel confident about the test coverage? I guess there's no test to ensure the success-counter works, but more importantly, do the put_item and get_item tests cover a reasonable portion of your code?

**Comment by ssiloti** [comment]
@arvidn actually neither the current code nor this patch has proper cas error handling. Currently a put that gets a cas error response will silently fail. With this patch dht_put_alert::num_success will less than it would have been, but the user can't even be sure that num_success is actually less than it should be, let alone that a cas error occurred.

**Comment by ssiloti** [comment]
Actually num_success won't even be reduced because put_data_observer doesn't check for an error response.

File: include/libtorrent/alert_types.hpp (hunk 18 above)
Comment (thomas-yuan) [inline include/libtorrent/alert_types.hpp]: I'll fix the comment.

No, current logic is, get k nodes, do put (same as announce). Even all failed, we won't try more nodes.

File: include/libtorrent/kademlia/dht_tracker.hpp (hunk 19 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/dht_tracker.hpp]: I didn't see the try again logic. The original code even can't report error to user because the dht_put_alert is sent before put.

File: include/libtorrent/kademlia/dht_tracker.hpp (hunk 20 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/dht_tracker.hpp]: I'll fix the comment.

File: include/libtorrent/kademlia/get_item.hpp (hunk 21 above)
Comment (thomas-yuan) [inline include/libtorrent/kademlia/get_item.hpp]: yes, same as announce/get_peers.

File: src/kademlia/get_item.cpp (hunk 24 above)
Comment (thomas-yuan) [inline src/kademlia/get_item.cpp]: The old logic is, we always call m_data_callback for immutable item, for get, it will sent dht_immutable_item_alert, and abort(), for put, it will get the data which going to put to DHT.

This patch add a new interface for put 
```void put_item(sha1_hash const& target, entry& data, boost::function<void(int)> f);```
so we needn't a callback to get data, as result, the m_data_callback is not set, and we can just return here because we don't care about what we get, we only need all nodes to put data.

File: src/kademlia/get_item.cpp (hunk 25 above)
Comment (thomas-yuan) [inline src/kademlia/get_item.cpp]: You are right. we couldn't assume other peer always doing right thing, maybe it's a attacker.

For the "powerful feature", what's the meaning of "update", don't put or put same value?
```
	bool put_immutable_item_callback(item& it, boost::function<void()> f)
	bool put_mutable_item_callback(item& it, bool authoritative, boost::function<void(item&)> cb)
```
in dht_tracker always return true, so if client call put_item, we will always do a PUT, even the blob is latest.

**Comment by thomas-yuan** [comment]
@ssiloti No, if response is error, it goes to time_out, then the num_success will be reduced.  But sure, as you said, user won't know what type errors there. (timeout / cas / invalid token etc).

File: src/kademlia/dht_tracker.cpp (hunk 23 above)
Comment (thomas-yuan) [inline src/kademlia/dht_tracker.cpp]: which interface should be simpler? dht_tracker::put_item? It's for put, so I guess it has separate get and put,  am I misreading this?

are you suggesting remove dht_tracker in this patch?

**Comment by thomas-yuan** [comment]
Thanks all comments here.

unit test is a issue.  since the alert include success_num now, it's a little bit hard to write unit test.
also get_item::got_data can be attacked is another issue.
 
working on these now.

File: src/kademlia/dht_tracker.cpp (hunk 22 above)
Comment (thomas-yuan) [inline src/kademlia/dht_tracker.cpp]: I think it actually related to the next comment - the dht_tracker is kinda redundant.
dht_tracker actually doesn't have any dht logic, for example it doesn't know traversal_algorithm, so I can't add the put_data logic in it. that said, node.cpp has to export a new interface for put.

**Comment by ssiloti** [comment]
@thomas-yuan you're right, I missed the bit of code in rpc_manager::incoming() which calls timeout() on any error response. This is going to need refactoring to be able to handle cas errors correctly.

**Comment by thomas-yuan** [comment]
rebase to master.
